sys/entropy_source/doc.txt

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (C) 2020 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+
+/**
+ * @defgroup     sys_entropy_source Entropy Sources
+ * @ingroup      sys
+ * @brief        Collection of different Entropy Sources
+ * @experimental This API is in an early state - expect changes.
+ * @warning      Entropy sources need to be thoroughly evaluated before deployment!
+ *
+ * @brief        Collection of entropy sources.
+ *
+ *
+ * This module adds support for additional entropy sources next to pure peripheral sources like
+ * @ref drivers_periph_hwrng and @ref sys_puf_sram. The concepts implemented here are heavily
+ * influenced by NIST SP 800-90B. Entropy sources can be used to feed more advanced entropy
+ * modules for cryptographic purposes, which typically accumulate multiple sources and safely
+ * maintain internal sates. Alternatively, these sources can be used directly or with internal
+ * conditioning enabled for non-cryptographic tasks like seed generation of general purpose PRNGs,
+ * in the absence of a hardware random number generator.
+ *
+ * Although APIs between multiple entropy sources might be similar, every source has to define its
+ * own API to be open for specific requirements. Entropy sources are crucial, thus, separate APIs should
+ * make the choice for one or the other explicit. Furthermore, multiple sources should be able to run in parallel
+ * and separate APIs simplify access (source management).
```

Comment:
> Sharing the same API has inherit merit, as it allows applications to easily make use of other implementations

This is not a user API, and I see only two use cases that involve entropy sources:
(i) High level entropy modules for cryptographic use that fetches values from multiple sources. Who cares if there are 2-4 software adapters? What is the "inhering merit" in practice?
(ii) A PRNG seeder (non crypto). Seeding is done by the OS (although I think the platform should be stronger involved).

> Note: Similar to SAUL

For the specific case that a developer really has to deal with entropy sources, I would rather argue the other way around. SAUL is a generic user interface that gives you "something" without caring what/where it comes from. That should be avoided when it comes to vulnerable entropy sources.


> An end user certainly doesn't want to deal with multiple different APIs, but a single high level API to access the already mixed entropy

Access to already mixed entropy is not the purpose of this PR.

Reply:
I agree that a common API in the case of entropy sources seems beneficial, and the distinction that @maribu makes between the types of entropy sources sounds reasonable. With this in mind, I also have to say, that I would not like for us to rush into an API that seems to cover the couple of use-cases that we can think of now, and fall into an early-generalization. I think a good example of API design has been the IEEE 802.15.4 radio HAL, which took into account multiple heterogeneous hardware and use-cases.

As I see it, introducing a common entropy source API would not invalidate the specific ADC noise API, which could be used to implement the common one. I propose that we move forward with this entropy source as is, leaving the design and evaluation of a common API for upcoming work. This way we can begin to benefit from this feature (and potentially others).

Reply:
Still, this whole paragraph should be removed and the API should be marked as internal. That way, refactoring the implementation to a common API can be done without maintaining multiple APIs.

In fact, IMO we should also mark the common entropy source API(s) as internal. I don't think that any application developer should ever directly interact with entropy sources, but instead interface with an entropy collection and mixing subsystem. Documenting the low level API(s) as internal might be one additional motivation for application programmers to do the right thing.

Reply:
@maribu, I have marked the APIs as `@internal` as you suggested, however, looking at it know I'd argue to remove it again. 
- Why hiding all information about this module? Even if an application developer would not use related APIs, it is worth to read the full documentation. Also for configuration purposes.
- System developers might need documentation anyway.
- The documentation has a warning and an experimental tag. The text describes how these APIs are (not) to be used. 
- The follow-up that ports an entropy module might make it even clearer which API to use (if at all, since application developers would probably not deal with entropy but a PRNG).

Reply:
I head more an `@note This API is considered as internal. Only use it if you know what you're doing and expect API changes without deprecation.` in mind.

Btw: The `@internal` is ignored by `make doc` due to the default Doxygen configuration. E.g. when looking at the [documentation of `STACK_CANARY_WORD`](https://api.riot-os.org/group__cpu__cortexm__common.html#ga2f0d4a98f077a68bc8e4046d80b9027b), there is no way to notice that this is intended to be internal.

Reply:
Thanks for the feedback. I have added a `@note` next to the `warning` and `experimental`. It should be very clear now.